### DISCUSSION TITLE
Getting a list of citekeys for html outputs

### DISCUSSION THREAD

**Discussion by zackbatist** [quarto-dev/quarto-cli#11734] [Q&A] [answered]
### Description

I know that it is possible to obtain an .aux file when generating PDFs by using `latex-clean: false`. But is there similar functionality to generate a list of citekeys actually used in a document when outputting to HTML only?

Related: Can citekeys be consolidated into a single aux file. For instance, when rendering a website containing multiple qmd files, it would be great to have an option to generate a single aux file for the entire website -- or perhaps this can be adapted to suit whatever target the yaml is configured for, i.e. if it's in a specific file's front matter, do it for that file, but if it's in a directory's _metadata.yml, then consolidate for all files covered by that.

**Comment by mcanouil** [comment]
Aux file is something specific to LaTeX.
What is it you are trying to do exactly? (Not how you think it should be done)

**Reply by mcanouil** [reply]
If you want to retrieve the list of citations actually cited regardless of the format, use a Lua filter.
- https://pandoc.org/lua-filters.html#type-cite
- https://quarto.org/docs/extensions/lua.html
- https://quarto.org/docs/extensions/lua-api.html
- https://quarto.org/docs/extensions/filters.html

**Reply by zackbatist** [reply]
I'm trying to obtain a list of citekeys referenced in a document or series of documents.

**Reply by zackbatist** [reply]
Thank you. I will reply if I manage to do this on my own or find a filter that already does this.

**Reply by mcanouil** [reply, marked as answer]
Something like that to write unique citation id into a file:

```lua
local citations = {}

function Cite(cite)
  for _, citation in ipairs(cite.citations) do
    citations[citation.id] = true
  end
end

function Pandoc(doc)
  local file = io.open("citekeys.txt", "w")
  for citekey, _ in pairs(citations) do
    file:write(citekey .. "\n")
  end
  file:close()
end
```

**Reply by zackbatist** [reply]
This is excellent, thank you. I wonder if this can be modified to collect all citekeys used throughout the project, rather than on a document-by-document basis.

**Reply by mcanouil** [reply]
Filters apply on document.
It can be applied on each document of a project.
So you can write the logic needed in the filter but you'll have to do that yourself.

**Reply by zackbatist** [reply]
Ok, thanks!
